[Unit Test] Create namespace parser unit test

j-ororke opened the pull request:
#### Summary

- Creating namespace parser with data provided by Cecille:
- Created new TestSpecParsingNamespace.py unit test to test namespaces
- Updated tests.yaml to include TestSpecParsingNamespace.py
- Updated test_metadata.yaml to include TestSpecParsingNamespace.py and reasoning
- Updated problem_notices support module to include new NamespacePathLocation function for Namespace file locations
- Updated spec_parsing to include new functions build_xml_namespaces and parse_namespace, as well as XmlNamespaces and XmlTags dataclasses

#### Related issues

Original PR that was split to allow easier reviews:
- PR Link: [38119](https://github.com/project-chip/connectedhomeip/pull/38119)

Provides unit tests for Cecille's Namespace parser PR:
- PR Link: [37527](https://github.com/project-chip/connectedhomeip/pull/37527)

#### Testing
- TestSpecParsingNamespace.py is a comprehensive unit test modules for validating Matter specification XML files versions 1.4 thru 1.5, it does some minor checking on 1.3 and master XML's but is not expected to be used for those XML files due to known issues with some of those at the time. It follows a two-phase testing approach: first testing with mock data to verify parser behavior with both valid and invalid inputs, then testing against actual specification XML files. The namespace parser performs extensive validation of namespace structures, ID formats, and version relationships, while maintaining proper error reporting, to the best of my knowledge this covered the requirement for the namespace parser request. Namespace parser unit test module ensures specification conformance while providing detailed problem reporting for any validation issues encountered.

j-ororke committed:
Creating TestSpecParsingNamespace test module:
- Creating with data provided by Cecille in PR here: https://github.com/project-chip/connectedhomeip/pull/37527
- Updated test.yaml to include TestSpecParsingNamespace.py to run in CI
- Updated matter_testing support module to include new NamespacePathLocation function for Namespace file locations
- Updated spec_parsing to include new functions build_xml_namespaces and parse_namespace, as well as XmlNamespaces and XmlTags dataclasses

j-ororke committed:
Updating test_metadata:
- Added TestSpecParsingNamespace.py as it is a unit test and not run against an app.

j-ororke committed:
Updating TestSpecParsingNamespace.py and spec_parsing.py:
- Updating to change method to using 0 instead of 16 bit enums
- Updating to use PrebuiltDatamodelDirectory instead of os to set filepaths
- Updated to utilizing the build_xml_namespaces function in spec_parsing.

j-ororke committed:
Update src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Updating TestSpecParsingNamespace and spec_parsing modules:
- Resolving linting errors

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating tests.yaml, spec_parsing.py, and constraints.txt:
- Updated to remove unneccessary apt install && apt update
- Updated to remove changing version of google-api in constraints.txt as no longer needed
- Updated to change commented out line method as better to have just # for single line entries and removed unneccesary __str__ function to format string before using it.

j-ororke committed:
Update constraints.txt

Removed commented out line 295 for prior google-api package version

j-ororke committed:
Updating TestSpecParsingNamespace python3 test modules:
- Added more comprehensive checks for namespaces between versions to TestSpecParsingNamespace
- Added known current differences between versions for namespaces
- Added change to contain only checks for versions 1.3, 1.4, 1.4.1, and 1.4.2, removed master as noticed in spec_parsing support module

j-ororke committed:
Updating TestSpecParsinNamespace unit test module:
- Minor changes

j-ororke committed:
Updating TestSpecParsingNamespace.py:
- Resolving stylizer issues: isort and autopep8

j-ororke committed:
autopep8 and isort fixes

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TestSpecParsingNamespace as locations of imports have changed recently:
- Added NamespacePathLocation dataclass function to problem_notices support module for correct import
- Changed to importing from problem_notices for NamespacePathLocation, ProblemNotice, ProblemSeverity to be used in TestSpecParsingNamespace module

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating namespace file check test as 'Common Closure' XML is now available for 1.4.2 release namespace

j-ororke commented on src/python_testing/TestSpecParsingNamespace.py:
```diff
@@ -0,0 +1,377 @@
+from chip.testing.problem_notices import NamespacePathLocation, ProblemNotice, ProblemSeverity
+from chip.testing.spec_parsing import (DataModelLevel, PrebuiltDataModelDirectory, build_xml_namespaces, get_data_model_directory,
+                                       parse_namespace)
+from jinja2 import Template
```
Thank you, this has been removed

j-ororke commented on src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py:
```diff
@@ -140,6 +142,37 @@ class XmlDeviceTypeClusterRequirements:
+@dataclass
+class XmlTag:
+    """Represents a tag within a namespace"""
+    def __init__(self):
+        self.id: int = 0
+        self.name: str = ""
+        self.description: Optional[str] = None
```
Updated!

j-ororke commented on src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py:
```diff
@@ -26,6 +26,7 @@
 from enum import Enum, auto
 from importlib.abc import Traversable
 from typing import Callable, Optional, Union
+import os
```
Thank you, removed!

j-ororke commented on src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py:
```diff
@@ -140,6 +142,37 @@ class XmlDeviceTypeClusterRequirements:
+@dataclass
+class XmlNamespace:
+    """Represents a namespace definition from XML"""
+    def __init__(self):
+        self.id: int = 0
+        self.name: str = ""
+        self.tags: dict[int, XmlTag] = {}
```
Thank you, updated!

j-ororke committed:
Namespace Parser Unit Test update:
- Removed unused imports in TestSpecParsingNamespace.py
- Removed redundant validate_namespace_xml function
- Simplified test_all_namespace_files to make it easier to maintain and cleaner, now relies on build_xml_namespaces functionality, this removed having to reparse and revalidate the data already validated earlier in the test module
- Enhanced the simplified test to make sure that all namespace files in 1.4.2 are being processed and validated.
- Removed the redundant explicit __init__ method from the NamespacePathLocation dataclass in problem_notices support module
- Removed the redundant explicit __init__ method from the XmlTag and XmlNamespace dataclasses in spec_parsing support module

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Adding back in the comprehensive namespace XML validation functionality

j-ororke committed:
Merge remote-tracking branch 'upstream/master' into create_namespace_parser_unittests

j-ororke committed:
Restoring NamespacePathLocation import in spec_parsing support module and chip -> matter namespace import changes

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Restoring UnknownProblemLocation import in spec_parsing support module

cecille reviewed (commented):
minor comments. I'd like to get someone else to take a look here too because I think some of this came from my branch, so it feels weird to be the only person approving it.

j-ororke committed:
Merge branch 'master' into create_namespace_parser_unittests

j-ororke commented on src/python_testing/matter_testing_infrastructure/matter/testing/spec_parsing.py:
```diff
@@ -799,6 +838,157 @@ def combine_attributes(base: dict[uint, XmlAttribute], derived: dict[uint, XmlAt
+
+    try:
+        # Handle both zip files and directories
+        if isinstance(namespace_dir, zipfile.Path):
```
Ah sorry, there is no real difference, I have replaced the isinstance() with the f.name.endswith('.xml') to match the other functions.

j-ororke commented on src/python_testing/TestSpecParsingNamespace.py:
```diff
@@ -0,0 +1,383 @@
+        one_four_one, one_four_one_problems = build_xml_namespaces(PrebuiltDataModelDirectory.k1_4_1)
+        one_four_two, one_four_two_problems = build_xml_namespaces(PrebuiltDataModelDirectory.k1_4_2)
+
+        asserts.assert_equal(len(one_four_problems), 0, "Problems found when parsing 1.4 spec")
```
I remember when we started working on the namespace parser there were noticed issues but it was expected at the time. However, I have added it back in now as I don't see any errors with it any longer.

j-ororke commented on src/python_testing/TestSpecParsingNamespace.py:
```diff
@@ -0,0 +1,383 @@
+
+        # Check changes from 1.3 to 1.4
+        removed_1_3_to_1_4 = set(one_three.keys()) - set(one_four.keys())
+        removed_names_1_3_to_1_4 = {one_three[id].name for id in removed_1_3_to_1_4}
```
Hi Cecille,
created issue here for this: https://github.com/project-chip/connectedhomeip/issues/40909

j-ororke committed:
Updating TestSpecParsingNamespace unit test and spec_parsing support modules:
- Changed to using for loop to gather XML directories for validating
- Added 1.5 XML as a directory to parse and compare prior XML directories with
- Readded validation of XML files for parsing 1.3 files
- Removed unneeded headers in spec_parsing support module
- changed from using isinstance() to using  f.name.endswith('.xml') in spec_parsing support module

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving lint

j-ororke committed:
Merge branch 'master' into create_namespace_parser_unittests

restyled-commits committed:
Restyled by isort

j-ororke committed:
Merge branch 'master' into create_namespace_parser_unittests

j-ororke committed:
Merge branch 'master' into create_namespace_parser_unittests

cecille reviewed (approved):
This looks reasonable to me, but I also wrote some of the initial code here, so I'm not comfortable being the sole approver. I'm going to pull in a second set of eyes to keep everything on the up and up.
